### PR TITLE
Fix: solve the convergence problem of E value in `dav_subspace` method （Useful to know how to avoid adding GlobalV）

### PR DESCRIPTION
**Pull request by haozhihan**
### Reminder
- [x] Have you linked an issue with this pull request?

### Linked Issue
#3447 

### Details
* In this PR, I added an input parameter called `diago_full_acc` referring to [Q-E's documentation](https://www.quantum-espresso.org/Doc/INPUT_PW.html#idm878).
* `diago_full_acc` can achieve a good balance between energy convergence accuracy and computational efficiency.

### CODE REVIEW

**Review by denghuilu — APPROVED** [review]

**Comment by haozhihan** [comment]
I will test more cases to prove that this PR is OK for coverage in `dav_subspace`. So, You can wait until I release the test results and then merge this PR.

**Comment by haozhihan** [comment]
I ran some more tests and got the results below. Please @WHUweiqingzhou and @pxlxingliang help me see if this convergence result is acceptable.

| cases | EDIFF(eV) in last scf | 
|---|---|
| AlMg_8 |  | 
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  |
| AlMg_8 |  || AlMg_8 |  |

**Comment by haozhihan** [comment]
I ran some more tests and got the results below. Please @WHUweiqingzhou and @pxlxingliang help me see if this convergence result is acceptable.

| cases | EDIFF(eV) in last scf | 
|---|---|
| AlMg_8 | -4.631041e-06 | 
| AlMg_12 | -7.404141e-06 |
| AlMg_16 | -2.779601e-06 |
| AlMg_20 | -4.197651e-06 |
| AlMg_24 | -2.312573e-07 |
| AlMg_40 | -9.019758e-07 |
| AlMg_48 | -1.821498e-07 |
| AlMg_pd04_8 | -2.425136e-06 |
| AlMg_pd04_12 | 1.218164e-08 |
| AlMg_pd04_16 | -6.655961e-06 |
| AlMg_pd04_20 | -2.506580e-06 |
| AlMg_pd04_24 | -8.934190e-07 |
| AlMg_pd04_32 | -7.040424e-07 |
| AlMg_pd04_40 | -3.453822e-06 |
| AlMg_pd04_48 | -3.094913e-06 |
| LiLaH_8 | -6.755438e-07 |
| LiLaH_12 | -7.632281e-07 |
| LiLaH_16 | -2.478116e-07 |
| LiLaH_20 | -1.349498e-06 |
| LiLaH_24 | -9.087260e-07 |
| LiLaH_32 | -1.052319e-06 |
| LiLaH_40 | -2.673772e-06 |

File: source/module_io/write_input.cpp
Comment (pxlxingliang) [inline source/module_io/write_input.cpp:116]: Please add the description of diago_full_acc in docs/advanced/input_files/input-main.md

Comment (haozhihan) [reply source/module_io/write_input.cpp:116]: I have added it in the latest commit.

**Comment by haozhihan** [comment]
I will retest whether the `EDIFF` of these cases is reasonable based on the latest commit code.

**Comment by haozhihan** [comment]
Update the result based on the last commit code.

| cases | EDIFF(eV) in last scf | 
|---|---|
| AlMg_8 | -6.046425e-06 | 
| AlMg_12 | -2.746880e-07 |
| AlMg_16 | -5.684387e-07 |
| AlMg_20 | -3.998047e-06 |
| AlMg_24 | -7.705483e-08 |
| AlMg_40 | -1.073471e-07 |
| AlMg_48 | -4.129307e-08 |
| AlMg_pd04_8 | -9.503761e-06 |
| AlMg_pd04_12 | -6.331262e-07 |
| AlMg_pd04_16 | -9.597795e-08 |
| AlMg_pd04_20 | -5.293841e-06 |
| AlMg_pd04_24 | -2.444094e-07 |
| AlMg_pd04_32 | -9.223745e-07 |
| AlMg_pd04_40 | -3.371806e-09 |
| AlMg_pd04_48 | 8.884871e-08 |
| LiLaH_8 | -1.194614e-07 |
| LiLaH_12 | -1.581466e-06 |
| LiLaH_16 | -1.643565e-06 |
| LiLaH_20 | -3.617514e-08 |
| LiLaH_24 | -8.098816e-07 |
| LiLaH_32 | 1.395165e-08 |
| LiLaH_40 | -8.427895e-07 |

We can see that the maximum EDIFF is approximately in the order of 1e-6. This is completely acceptable.

The above results are calculated under the default condition of `diago_full_acc=false`.

When `diago_full_acc=true`, higher precision calculation results will be obtained.

File: source/module_base/global_variable.cpp
Comment (mohanchen) [inline source/module_base/global_variable.cpp]: do not add new global variable

File: source/module_base/global_variable.h
Comment (mohanchen) [inline source/module_base/global_variable.h]: stop adding new global variable

File: source/module_hsolver/diago_dav_subspace.h
Comment (mohanchen) [inline source/module_hsolver/diago_dav_subspace.h]: changes DAV_BIG_THR to dav_big_thr, and what's the meaning of 'big'? typically we use large, not big

File: source/module_hsolver/hsolver_pw.cpp
Comment (mohanchen) [inline source/module_hsolver/hsolver_pw.cpp]: stop using GlobalV

File: docs/advanced/input_files/input-main.md
Comment (mohanchen) [inline docs/advanced/input_files/input-main.md]: could you write down the exact value of the threshold?

File: source/module_hsolver/diago_dav_subspace.cpp
Comment (mohanchen) [inline source/module_hsolver/diago_dav_subspace.cpp]: Is this a const double?

File: source/module_hsolver/diago_dav_subspace.h
Comment (mohanchen) [inline source/module_hsolver/diago_dav_subspace.h:34]: How about initialize the variable?

File: source/module_hsolver/hsolver_pw.cpp
Comment (mohanchen) [inline source/module_hsolver/hsolver_pw.cpp:179]: add a check in case wk[i]=0.0, which will cause segmentation fault.

File: source/module_hsolver/hsolver_pw.h
Comment (mohanchen) [inline source/module_hsolver/hsolver_pw.h:30]: add a doxygen style explanation for this variable

File: source/module_hsolver/hsolver_pw.cpp
Comment (haozhihan) [reply source/module_hsolver/hsolver_pw.cpp:179]: look line175: if (pes->klist->wk[i] > 0.0)

File: source/module_hsolver/diago_dav_subspace.h
Comment (haozhihan) [reply source/module_hsolver/diago_dav_subspace.h:34]: look line143: double Diago_DavSubspace<Real, Device>::dav_large_thr = 1e-5;

File: source/module_hsolver/diago_dav_subspace.cpp
Comment (haozhihan) [reply source/module_hsolver/diago_dav_subspace.cpp]: yes

Comment (mohanchen) [reply source/module_hsolver/diago_dav_subspace.cpp]: then modify it

**Review by mohanchen — APPROVED** [review]
LGTM

**Comment by mohanchen** [comment]
This PR shows how to avoid using GlobalV to add new input parameters.

**Pull request merged**
